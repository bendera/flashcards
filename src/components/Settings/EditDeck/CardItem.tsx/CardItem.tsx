import { ChangeEvent, FC, useState } from 'react';
import { Button, FormGroup, InputGroup, TextArea } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { FlashCard } from 'types';
import noop from 'utils/noop';
import styles from './CardItem.module.css';

interface CardItemProps {
  card: FlashCard;
  onChange?: (card: FlashCard) => void;
  onDelete?: (id: string) => void;
  onSwap?: (card: FlashCard) => void;
}

const CardItem: FC<CardItemProps> = ({
  card,
  onChange = noop,
  onDelete = noop,
  onSwap = noop,
}) => {
  const { id, frontSide, backSide } = card;

  const handleChange = (event: any) => {
    const el = event.currentTarget as HTMLInputElement;
    const val = el.value;
    const payload: FlashCard = {
      id,
      frontSide,
      backSide,
    };

    switch (el.id) {
      case `f_${id}`:
        payload.frontSide = val;
        break;
      case `b_${id}`:
        payload.backSide = val;
        break;
    }

    onChange(payload);
  };

  const handleDelete = () => {
    onDelete(id);
  };

  const handleSwap = () => {
    onSwap(card);
  };

  return (
    <div className={styles.wrapper}>
      <FormGroup
        className={styles.formGroup}
        label="Front side"
        labelFor={`f_${id}`}
      >
        {/* <InputGroup
          id={`f_${id}`}
          onInput={handleChange}
          value={frontSide}
        /> */}
        <TextArea
          id={`f_${id}`}
          onChange={handleChange}
          value={frontSide}
          fill
          growVertically
          className={styles.textarea}
        />
      </FormGroup>
      <Button
        className={styles.swapButton}
        icon={IconNames.SWAP_HORIZONTAL}
        minimal
        onClick={handleSwap}
      ></Button>
      <FormGroup
        className={styles.formGroup}
        label="Back side"
        labelFor={`b_${id}`}
      >
        <InputGroup id={`b_${id}`} onInput={handleChange} value={backSide} />
      </FormGroup>
      <Button
        className={styles.deleteButton}
        icon={IconNames.TRASH}
        minimal
        onClick={handleDelete}
      ></Button>
    </div>
  );
};

export default CardItem;
