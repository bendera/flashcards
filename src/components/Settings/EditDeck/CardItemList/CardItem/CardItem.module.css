.root {
  align-items: flex-start;
  display: flex;
  width: 100%;
}

.root .formGroup {
  margin-bottom: 0;
  width: 100%;
}

.checkbox {
  margin-top: 22px;
}

.swapButton,
.deleteButton {
  margin-top: 22px;
}

.swapButton {
  margin-left: 4px;
  margin-right: 4px;
}

.deleteButton {
  margin-left: 4px;
}

.textarea {
  resize: none;
}