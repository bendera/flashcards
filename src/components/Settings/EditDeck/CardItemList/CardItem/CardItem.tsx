import { ChangeEvent, FC, FormEvent } from 'react';
import { Button, Checkbox, FormGroup, TextArea } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import cn from 'classnames';
import noop from 'utils/noop';
import { CardListItemData } from '../CardItemList';
import styles from './CardItem.module.css';

interface CardItemProps {
  card: CardListItemData;
  className?: string;
  onChange?: (item: CardListItemData) => void;
  onDelete?: (card: CardListItemData) => void;
}

const CardItem: FC<CardItemProps> = ({
  card,
  className,
  onChange = noop,
  onDelete = noop,
}) => {
  const { id, frontSide, backSide, selected } = card;

  const handleCheckboxChange = (event: FormEvent<HTMLInputElement>) => {
    const target = event.target as HTMLInputElement;

    if (target.type === 'checkbox') {
      onChange({
        id,
        frontSide,
        backSide,
        selected: target.checked,
      });
    }
  };

  const handleChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
    const el = event.currentTarget as HTMLTextAreaElement;
    const val = el.value;
    const payload: CardListItemData = {
      id,
      frontSide,
      backSide,
      selected,
    };

    switch (el.id) {
      case `f_${id}`:
        payload.frontSide = val;
        break;
      case `b_${id}`:
        payload.backSide = val;
        break;
    }

    onChange(payload);
  };

  const handleDelete = () => {
    onDelete(card);
  };

  const handleSwap = () => {
    onChange({
      id,
      frontSide: backSide,
      backSide: frontSide,
      selected,
    });
  };

  return (
    <div className={cn(styles.root, className)}>
      <Checkbox
        checked={selected}
        className={styles.checkbox}
        data-testid="CardItem__checkbox"
        onChange={handleCheckboxChange}
        value={id}
      />
      <FormGroup
        className={styles.formGroup}
        label="Front side"
        labelFor={`f_${id}`}
      >
        <TextArea
          className={styles.textarea}
          fill
          growVertically
          id={`f_${id}`}
          onChange={handleChange}
          defaultValue={frontSide}
        />
      </FormGroup>
      <Button
        className={styles.swapButton}
        icon={IconNames.SWAP_HORIZONTAL}
        minimal
        onClick={handleSwap}
        title="Swap sides"
      ></Button>
      <FormGroup
        className={styles.formGroup}
        label="Back side"
        labelFor={`b_${id}`}
      >
        <TextArea
          id={`b_${id}`}
          growVertically
          fill
          onChange={handleChange}
          defaultValue={backSide}
          className={styles.textarea}
        />
      </FormGroup>
      <Button
        className={styles.deleteButton}
        icon={IconNames.TRASH}
        minimal
        onClick={handleDelete}
        title="Delete card"
      ></Button>
    </div>
  );
};

export default CardItem;
