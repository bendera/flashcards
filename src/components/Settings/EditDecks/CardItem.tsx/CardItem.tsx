import { FC } from 'react';
import { Button, FormGroup, InputGroup } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { FlashCard } from 'types';
import styles from './CardItem.module.css';

interface CardItemProps {
  card: FlashCard;
}

const CardItem: FC<CardItemProps> = ({ card }) => {
  const { id, frontSide, backSide } = card;

  return (
    <div className={styles.wrapper}>
      <FormGroup
        labelFor={`f_${id}`}
        label="Front side"
        className={styles.formGroup}
      >
        <InputGroup id={`f_${id}`} />
      </FormGroup>
      <Button
        icon={IconNames.SWAP_HORIZONTAL}
        minimal
        className={styles.swapButton}
      ></Button>
      <FormGroup
        labelFor={`b_${id}`}
        label="Back side"
        className={styles.formGroup}
      >
        <InputGroup id={`b_${id}`} />
      </FormGroup>
      <Button
        icon={IconNames.TRASH}
        minimal
        className={styles.deleteButton}
      ></Button>
    </div>
  );
};

export default CardItem;
