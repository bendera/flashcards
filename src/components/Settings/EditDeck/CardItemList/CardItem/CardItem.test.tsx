import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CardItem from './CardItem';

describe('CardItem', () => {
  it('should render correctly with required props', () => {
    const { queryByTestId, container } = render(
      <CardItem
        card={{
          id: 'testId',
          frontSide: 'Test front',
          backSide: 'Test back',
          selected: true,
        }}
      />
    );

    const cb = queryByTestId('CardItem__checkbox') as HTMLInputElement;

    expect(cb?.checked).toBe(true);
    expect(container).toMatchSnapshot();
  });

  it('onChange should be called with proper params when checkbox is changed', () => {
    const changeHandler = jest.fn();

    const { queryByTestId } = render(
      <CardItem
        card={{
          id: 'testId',
          frontSide: 'Test front',
          backSide: 'Test back',
          selected: false,
        }}
        onChange={changeHandler}
      />
    );

    const cb = queryByTestId('CardItem__checkbox');
    cb!.click();

    expect(changeHandler).toHaveBeenCalledWith({
      backSide: 'Test back',
      frontSide: 'Test front',
      id: 'testId',
      selected: true,
    });
  });

  it('onChange should be called with proper params when sides are swapped', () => {
    const changeHandler = jest.fn();

    const { queryByTitle } = render(
      <CardItem
        card={{
          id: 'testId',
          frontSide: 'Test front',
          backSide: 'Test back',
          selected: false,
        }}
        onChange={changeHandler}
      />
    );

    const bt = queryByTitle('Swap sides');
    bt!.click();

    expect(changeHandler).toHaveBeenCalledWith({
      backSide: 'Test front',
      frontSide: 'Test back',
      id: 'testId',
      selected: false,
    });
  });

  it('onChange should be called with proper params when value of the front side textarea changed', async () => {
    const changeHandler = jest.fn();
    const user = userEvent.setup();
    const TYPED_TEXT = 'test';

    const { queryAllByRole } = render(
      <CardItem
        card={{
          id: 'testId',
          frontSide: '',
          backSide: 'Test back',
          selected: false,
        }}
        onChange={changeHandler}
      />
    );

    const ta = queryAllByRole('textbox')[0];
    await user.type(ta, TYPED_TEXT);

    expect(changeHandler).toHaveBeenCalledTimes(TYPED_TEXT.length);
    expect(changeHandler).toHaveBeenLastCalledWith({
      backSide: 'Test back',
      frontSide: 'test',
      id: 'testId',
      selected: false,
    });
  });

  it('onChange should be called with proper params when value of the back side textarea changed', async () => {
    const changeHandler = jest.fn();
    const user = userEvent.setup();
    const TYPED_TEXT = 'test';

    const { queryAllByRole } = render(
      <CardItem
        card={{
          id: 'testId',
          frontSide: 'Test front',
          backSide: '',
          selected: false,
        }}
        onChange={changeHandler}
      />
    );

    const textarea = queryAllByRole('textbox')[1];
    await user.type(textarea, TYPED_TEXT);

    expect(changeHandler).toHaveBeenCalledTimes(TYPED_TEXT.length);
    expect(changeHandler).toHaveBeenLastCalledWith({
      id: 'testId',
      frontSide: 'Test front',
      backSide: TYPED_TEXT,
      selected: false,
    });
  });

  it('onDelete should be called with proper params when card is removed', () => {
    const deleteHandler = jest.fn();

    const { queryByTitle } = render(
      <CardItem
        card={{
          id: 'testId',
          frontSide: 'Test front',
          backSide: 'Test back',
          selected: false,
        }}
        onDelete={deleteHandler}
      />
    );

    const bt = queryByTitle('Delete card');
    bt!.click();

    expect(deleteHandler).toHaveBeenCalledWith({
      backSide: 'Test back',
      frontSide: 'Test front',
      id: 'testId',
      selected: false,
    });
  });
});
