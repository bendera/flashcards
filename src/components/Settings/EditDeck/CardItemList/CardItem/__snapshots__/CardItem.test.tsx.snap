// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItem should render correctly with required props 1`] = `
<div>
  <div
    class="root"
  >
    <label
      class="bp4-control bp4-checkbox checkbox"
    >
      <input
        checked=""
        data-testid="CardItem__checkbox"
        type="checkbox"
        value="testId"
      />
      <span
        class="bp4-control-indicator"
      />
    </label>
    <div
      class="bp4-form-group formGroup"
    >
      <label
        class="bp4-label"
        for="f_testId"
      >
        Front side
         
        <span
          class="bp4-text-muted"
        />
      </label>
      <div
        class="bp4-form-content"
      >
        <textarea
          class="bp4-input bp4-fill textarea"
          id="f_testId"
          style="height: 0px;"
        >
          Test front
        </textarea>
      </div>
    </div>
    <button
      class="bp4-button bp4-minimal swapButton"
      title="Swap sides"
      type="button"
    >
      <span
        aria-hidden="true"
        class="bp4-icon bp4-icon-swap-horizontal"
        icon="swap-horizontal"
      >
        <svg
          data-icon="swap-horizontal"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M0 7.02L.05 7H0v.02zm2-2.03h9.57l-1.29 1.29A1.003 1.003 0 0011.7 7.7l2.99-3c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71l-2.99-3a1.07 1.07 0 00-.71-.28 1.003 1.003 0 00-.71 1.71L11.57 3H2c-.55 0-1 .45-1 1a1 1 0 001 .99zM15.96 9H16v-.02l-.04.02zM14 11.01H4.43l1.29-1.29A1.003 1.003 0 004.3 8.3l-2.99 3a.99.99 0 00-.29.7c0 .28.11.53.29.71l2.99 3a1.003 1.003 0 001.42-1.42L4.43 13H14c.55 0 1-.45 1-1s-.45-.99-1-.99z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </button>
    <div
      class="bp4-form-group formGroup"
    >
      <label
        class="bp4-label"
        for="b_testId"
      >
        Back side
         
        <span
          class="bp4-text-muted"
        />
      </label>
      <div
        class="bp4-form-content"
      >
        <textarea
          class="bp4-input bp4-fill textarea"
          id="b_testId"
          style="height: 0px;"
        >
          Test back
        </textarea>
      </div>
    </div>
    <button
      class="bp4-button bp4-minimal deleteButton"
      title="Delete card"
      type="button"
    >
      <span
        aria-hidden="true"
        class="bp4-icon bp4-icon-trash"
        icon="trash"
      >
        <svg
          data-icon="trash"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M14.49 3.99h-13c-.28 0-.5.22-.5.5s.22.5.5.5h.5v10c0 .55.45 1 1 1h10c.55 0 1-.45 1-1v-10h.5c.28 0 .5-.22.5-.5s-.22-.5-.5-.5zm-8.5 9c0 .55-.45 1-1 1s-1-.45-1-1v-6c0-.55.45-1 1-1s1 .45 1 1v6zm3 0c0 .55-.45 1-1 1s-1-.45-1-1v-6c0-.55.45-1 1-1s1 .45 1 1v6zm3 0c0 .55-.45 1-1 1s-1-.45-1-1v-6c0-.55.45-1 1-1s1 .45 1 1v6zm2-12h-4c0-.55-.45-1-1-1h-2c-.55 0-1 .45-1 1h-4c-.55 0-1 .45-1 1v1h14v-1c0-.55-.45-1-1-1z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
